### PR TITLE
Update to hatch-based build system and fix CI

### PR DESCRIPTION
**Pull request by ianthomas23**
This PR switches the build system from `jupyter_packaging` to the recommended `hatch_jupyter_builder`, plus some other smaller changes, in an attempt to get CI passing. When it succeeds and this PR is merged, I will deal with other improvements such as bringing dependencies up to date.

Summary of changes made:
- Switch to `hatch`-based builder.
- Update `micromamba` github actions.
- Use `python -m build` to build the package rather than the deprecated `python setup.py ...`
- Use `jlpm` rather than `yarn` for TypeScript package management and building.

With these changes I can build and run the tests locally, and also install from the generated sdist and binary wheel.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,7 +1,98 @@
...
+    "Framework :: IPython",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Science/Research",
+    "Programming Language :: Python :: 3.5",
+    "Programming Language :: Python :: 3.6",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
```
Comment (martinRenou) [inline pyproject.toml:33]: Should we add Python 3.12 here?

Comment (ianthomas23) [reply pyproject.toml:33]: Yes, I don't know why I stopped at 3.11.

We should probably decide on what Python versions we are supporting, heading toward [SPEC0](https://scientific-python.org/specs/spec-0000/) eventually. Maybe 3.9 to 3.12 is appropriate now. Then I could ensure that CI is running using both the earliest and latest Python versions that we claim to support.

**Comment by ianthomas23** [comment]
> Would be nice to setup the jupyter-releaser after that

I'll take a look at jupyter-releaser.

File: pyproject.toml (hunk 1 above)
Comment (martinRenou) [reply pyproject.toml:33]: Sounds good 👍🏽

**Comment by martinRenou** [comment]
Bot update galata references

**Comment by martinRenou** [comment]
The bot is failing for some other reason. Anyway I believe the CI failures are unrelated. It seems the visual regression diff is super small and probably due to some update on the matplotlib side. I can look into getting this green separately.

**Comment by ianthomas23** [comment]
I am going to look at excluding text from the Matplotlib plots. This is done in Matplotlib CI as there can be tiny differences depending on platform, OS and compiler.

**Comment by martinRenou** [comment]
Thank you! I guess this PR is good to go as-is!

> Fix update_galata_references.yaml

The fix could probably be just https://github.com/jupytercad/jupytercad/blob/main/.github/workflows/update_galata_references.yaml#L31

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -49,25 +46,28 @@ jobs:
...
         include:
-          - python: '3.10'
-            dist: 'ipympl*.tar.gz'
+          # Earliest supported Matplotlib, needs python <= 3.10
           - python: '3.9'
+            mpl-version: '3.5'
             dist: 'ipympl*.whl'
+          # Install from sdist not wheel
+          - python: '3.12'
+            mpl-version: 'latest'
```
Comment (martinRenou) [inline .github/workflows/main.yml:59]: Are we missing an `os` entry here? I wonder if this is the reason why the workflow does not start

Comment (ianthomas23) [reply .github/workflows/main.yml:61]: Yes, I've corrected this now.

Could you cancel the old CI runs as I don't have permissions to do that?

**Comment by ianthomas23** [comment]
I changed the test that installed from sdist from python 3.12 to 3.11 so now it passes. With python 3.12 it was trying to compile zmq from source. I don't think that should be necessary, but it is not a problem for us to solve today.

File: .github/workflows/main.yml (hunk 2 above)
Comment (ianthomas23) [reply .github/workflows/main.yml:61]: Thanks!

**Comment by martinRenou** [comment]
> I changed the test that installed from sdist from python 3.12 to 3.11 so now it passes. With python 3.12 it was trying to compile zmq from source. I don't think that should be necessary, but it is not a problem for us to solve today.

We may be able to workaround this issue by installing from source with `--no-deps` and making sure all dependencies are installed using conda/mamba from conda-forge.

Thank you for getting this green! Merging and we can iterate

**Review by martinRenou — APPROVED** [review]
Thank you!

**Pull request merged**